### DISCUSSION TITLE
How does search work with &nbsp;?

### DISCUSSION THREAD

**Discussion by Aster89** [qutebrowser/qutebrowser#7477] [Q&A] [answered]
This is mostly a curiosity.

I was reading [this](https://sysprog21.github.io/lkmpg/#chardevc), and at some point I searched for the string _Creates a read-only_ (with an upper case _C_), but it found no result on the page.

![notfound](https://user-images.githubusercontent.com/20521900/201516168-8f5c63a6-b65d-455f-9359-78863c4b7cf4.png)

Then I removed the leading _C_, thus searching for _reates a read-only_, and match was found indeed.

![spaces](https://user-images.githubusercontent.com/20521900/201516053-b57a6380-e88a-4109-a068-f407aa7375bf.png)

I don't really understand why those `&nbsp;` are there; I mean, I know "nbsp" stands for "non-breakable space", but I'm a bit puzzled as to what happens if I double click on the content of the element in the chrome dev tools:
![doubleclick](https://user-images.githubusercontent.com/20521900/201516104-2247d99c-48ac-496d-afa5-f117e7876803.png)

P.S.: Chromium does find it.

**Comment by toofar** [comment, marked as answer]
What's your `search.ignore_case` set to? For me, and the default, it is set to `smart`, which means if you have a capital letter in the search string it does a case sensitive search, an all lower-case search string will do a case insensitive search.
For me on that page searching for the full `creates a read-only` in lower case does find it, changing the C to upper case makes it not find any matches though, like you saw.
So it seems the case sensitive search gets tripped up on the non-breaking spaces, possibly it doesn't treat all whitespaces the same anymore. We are just using the [WebEngine implementation](https://doc.qt.io/qt-6/qwebenginepage.html#FindFlag-enum) here, I have no idea if this is intentional by them or not.

**Reply by Aster89** [reply]
I confirm the exact setting and behavior you describe.

**Reply by The-Compiler** [reply]
So, QtWebEngine does just pass the "case (in)sensitive" flag to Chromium, which passes it through some Chromium layers, which then passes it to some Blink layers, and [finally it ends up here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/editing/iterators/text_searcher_icu.cc;l=233-234;drc=49f9d9415a5616b6061daa80eb3c737f5e2e8253), setting ICUs `const UCollationStrength strength = case_sensitive ? UCOL_TERTIARY : UCOL_PRIMARY;`.

The [ICU docs say](https://github.com/unicode-org/icu/blob/main/docs/userguide/collation/concepts.md):

> Primary Level: Typically, this is used to denote differences between base characters (for example, "a" < "b"). It is the strongest difference. For example, dictionaries are divided into different sections by base character. This is also called the level-1 strength.
>
> Secondary Level: Accents in the characters are considered secondary differences (for example, "as" < "às" < "at"). Other differences between letters can also be considered secondary differences, depending on the language. A secondary difference is ignored when there is a primary difference anywhere in the strings. This is also called the level-2 strength. Note: In some languages (such as Danish), certain accented letters are considered to be separate base characters. In most languages, however, an accented letter only has a secondary difference from the unaccented version of that letter.
>
> Tertiary Level: Upper and lower case differences in characters are distinguished at the tertiary level (for example, "ao" < "Ao" < "aò"). In addition, a variant of a letter differs from the base form on the tertiary level (such as "A" and "Ⓐ"). Another example is the difference between large and small Kana. A tertiary difference is ignored when there is a primary or secondary difference anywhere in the strings. This is also called the level-3 strength.

Thus, case-sensitivity (level 3) always also includes normalizing accents (level 2). Given that a non-breaking space [decomposes into a normal space](https://www.fileformat.info/info/unicode/char/00a0/index.htm), I suppose those are also considered equal with case-insensitivity, but not with case-sensitivity. It's like searching for `bä` matching `Ba` too, but searching for `Bä` won't.

Technically this is a Blink bug (or at least API weirdness), but given that Chromium [doesn't even expose case-insensitive search](https://bugs.chromium.org/p/chromium/issues/detail?id=687552), it's not like anyone would care...
